### DISCUSSION TITLE
GHAS Certification Exam Prep: Part One - Getting Started 🔒

### DISCUSSION THREAD

**Discussion by queenofcorgis** [community/community#137620] [Enterprise]
_This discussion and  some of its comments have been edited and modified from its original format to enable evergreen learning_

Welcome to the first part of the [GitHub Advanced Security certification prep course](https://github.com/orgs/community/discussions/137197)! To begin, we’ll be going over some intro materials and best practices to build a strong foundation for the exam. This discussion is your space to ask questions, review study guides, and do some exam practice. 

## Step One: Prep 📚

We’ve assembled some materials for this first section. 

* Microsoft Learn Modules
   * [Introduction to GitHub Advanced Security](https://learn.microsoft.com/en-us/training/modules/introduction-github-advanced-security/?ns-enrollment-type=Collection&ns-enrollment-id=rqymc6yw8q5rey)
    * [Configure Dependabot security updates on your GitHub repo ](https://learn.microsoft.com/en-us/training/modules/configure-dependabot-security-updates-on-github-repo/?ns-enrollment-type=Collection&ns-enrollment-id=rqymc6yw8q5rey)
* Study guides from @nickliffen 
   * [Understanding GitHub Advanced Security](https://resources.github.com/learn/pathways/security/essentials/application-security-testing-github-advanced-security)
   * [Enabling GitHub Advanced Security](https://resources.github.com/learn/pathways/security/essentials/enabling-github-advanced-security)
* Study guides from @ladykerr 
   * [Describe the GHAS features and their functionality](https://github.com/LadyKerr/github-certification-guide/blob/main/study-guides/gh-advanced-security.md#domain-1-describe-the-ghas-security-features-and-functionality) 
   * [Configure and use Dependency Management ](https://github.com/LadyKerr/github-certification-guide/blob/main/study-guides/gh-advanced-security.md#domain-3-configure-and-use-dependency-management)
* Videos 
   * GitHub Advanced Security - Pass the Exam from freeCodeCamp
      * [GHAS Overview](https://www.youtube.com/watch?v=i740xlsqxEM&t=0s)
      * [Dependency Management](https://www.youtube.com/watch?v=i740xlsqxEM&t=3677s)
* Community member materials
  * @Salgado2004's [breakdown in their comment](https://github.com/orgs/community/discussions/137620#discussioncomment-10578129)

### Question 1: Which of the following is a feature of Dependabot?
A) Dependabot automatically commits changes to the main branch.
B) Dependabot opens pull requests to update dependencies.
C) Dependabot only works with public repositories.
D) Dependabot can update Docker images directly.

### Question 2: Which of the following features is part of GitHub Advanced Security?
A) Secret Scanning to detect and alert on leaked secrets in your code.
B) Dependabot to create alerts on and update insecure dependencies.
C) Code Scanning to identify and fix security vulnerabilities and other errors in your code.
D) All of the above

### Question 3: Which of the following is a feature of GitHub Advanced Security?
A) Real-time collaboration.
B) Automated code formatting.
C) Dependency scanning.
D) Issue tracking.

_[View the answers in my comment](https://github.com/orgs/community/discussions/137620#discussioncomment-10572841) 🧠_

**Use the discussion below to share additional study resources and respond to our prep questions**

[Join Part Two's Discussion on Scanning](https://github.com/orgs/community/discussions/138220)

**Comment by ghost** [comment]
Awesome curated list of resources! For the Answers, I think,

1. Option B
2. Option D
3. Option C

**Reply by hubertassogbavi6** [reply]
B

**Comment by juanlujand** [comment]
Question 1: Which of the following is a feature of Dependabot?
B) Dependabot opens pull requests to update dependencies.

Question 2: Which of the following features is part of GitHub Advanced Security?
D) All of the above

Question 3: Which of the following is a feature of GitHub Advanced Security?
C) Dependency scanning.

**Comment by Yordan-Ramchev** [comment]
1. B
2. D
3. C

**Comment by bos-hieu** [comment]
1.B
2.D
3.C

**Reply by keerthan0521** [reply]
1.B
2.D
3.C

**Reply by keerthan0521** [reply]
Thank you for doing this. I have completed my github action cert and currently preparing advanced security cert. Again thanks for awesome curated list of resources! 
Answers: 
Option B
Option D
Option C

**Comment by KeorapetseMalgas** [comment]
1.B
2.D
3.C

**Comment by amalayverma** [comment]
B, D, C

**Comment by Jcodebroken** [comment]
**1.** b
**2.** d
**3.** c

**Comment by Rei-Suzuki1729** [comment]
1. B
2. D
3. C

**Comment by queenofcorgis** [comment]
The answers are...

<details>
<summary>Answers</summary>
Question 1: Which of the following is a feature of Dependabot?
Correct Answer: B) Dependabot opens pull requests to update dependencies.
Explanation:
A) Incorrect. Dependabot opens pull requests rather than directly committing changes to the main branch.
B) Correct. Dependabot's primary feature is to open pull requests to update dependencies.
C) Incorrect. Dependabot supports both public and private repositories.
D) Incorrect. Dependabot does not support updating Docker images directly.


Question 2: Which of the following features is part of GitHub Advanced Security?
All above, to read more about GHAS's features, check out [this doc](https://docs.github.com/en/get-started/learning-about-github/about-github-advanced-security#about-advanced-security-features). 

Question 3: Which of the following is a feature of GitHub Advanced Security?
Correct Answer: C) Dependency scanning, 
</details>

**Reply by Salgado2004** [reply]
I have some!
- Does the GHAS license is included in GitHub Enterprise?
- Does Code scanning provide IDE extensions?

**Reply by samsmithnz** [reply]
The GHAS license is separate from the GitHub Enterprise license. https://github.com/enterprise/advanced-security#pricing

Can you provide a little more information about the IDE extension you are looking for with Code Scanning?

**Reply by soumya03Surya** [reply]
There is [CodeQL Extension](https://marketplace.visualstudio.com/items?itemName=GitHub.vscode-codeql) for VS Code - https://docs.github.com/en/enterprise-cloud@latest/code-security/codeql-for-vs-code/getting-started-with-codeql-for-vs-code

**Comment by sulemanxo** [comment]
1. B
2. D
3. C

**Comment by Salgado2004** [comment]
I'm a little late for this weeks lesson, but let me contribute!

Here's my breakdown for this topic, based on GHAS study guide 😅 I've left some notes with my personal questions, feel free to complement my breakdown and answer my questions, please! I loved the initiative and I'm greateful for the list of resources! Certainly they're going to help 🚀

## Domain 1 : Describe the GHAS security features and functionality
### Contrast GHAS features and their role in the security ecosystem
- **Differentiate the security features that come automatically for open source projects, and what features are available when GHAS is paired with GHEC or GHES**:
	Open source projects and **public repositories in general** automatically have secret scanning and a dependency graph, and can also enable code scanning and dependabot, but as a **foundational** resource. With the GitHub Advanced Security license, wich provide a **deeper protection** that is needed for **complex and enterprise projects**. Those additional features include secret scanning, code scanning and dependency review for private repos.
> [!note]
> I didn't get the GHES and GHEC part. Do you need the GitHub Enterprise to be able to purchase the GHAS license? Does the license is included in GitHub Enterprise? 
- **Describe the features and benefits of Security Overview**:
	Security Overview is a feature for GitHub Enterprise. It enables managing repositories from your organization and see which repositories have enabled specific security features. It also allows you to configure available security resources that are not currently in use (like a checklist?). 
- **Describe the differences between secret scanning and code scanning**:
	**Secret scanning** is a feature that analyses commits for **potencial leaked secrets** (such as API keys or secret tokens) based on predefined patterns of sensitive information. **Code scanning** is a **SAST tool** (such as Fortify or SonarQube) to statically analyses your code and look for vulnerabilities, like SQL injection or XSS. 
> [!note] 
> Does Code scanning provide IDE extensions? Like VS Code integration or something (would be cool)
- **Describe how secret scanning, code scanning, and Dependabot create a more secure software development life cycle**: 
	I believe that these three features **complete each other**, in the way that each one takes care of one kind of **security concerns**. By enabling all of them, you can manage and implement different security compliant features directly in your **development environment**.
- **Contrast a security scenario with isolated security review and an advanced scenario, with security integrated into each step of the software development life cycle**:
	I don't have much experience in this area... but I could say that isolated security reviews certainly are not as useful, specially because they are usually done _after_ the development cycle (or in the final stages) so you have a lot o rework or refactor to do, sometimes in **critical points of the application** that would affect **how the software was constructed**. An advanced scenario with security steps integrated, you can manage security concerns in each commit, **avoid great changes** in the future, the so called "Early Detection". Which is much better.
### Explain and use specific GHAS features
- **Describe how vulnerable dependencies are identified (by looking at the manifest files and comparing with databases of known vulnerabilities)**:
	This is a work fot dependabot 🤖! It is an automated tool for detecting vulnerable dependencies, cause it analyses the explicit declared dependencies in manifest and lockfiles in your repository (which fortunatelly come with the version of the dependency) and search them in **GitHub Advisory Database**, which is a security vulnerability database that you can also manually check if you are interested.
- **Explain how to act on alerts from GHAS**:
	To see and manage your alerts, you can go to the **Security tab** of the repository. There you'll find all the alerts for the repository and you can choose to **ignore them or fix them**. Before taking one decision, it is important to **investigate the nature and severity** of the alert. If you choose to fix, GitHub has some **automation to help you** (at least for dependabot!).
> [!note]
> The only alert that I have enabled so far are the dependabot alerts. I have yet to learn how to use the other ones 😅 but it seems pretty cool
- **Explain the implications of ignoring an alert**:
	It can led to a lot of risks, correct? Ignoring an alert means that your software has potencial vulnerabilities that could be exploited with malicious intentions. 
- **Explain the role of a developer when they discover a security alert**:
	As a developer, therefore the maintainer of the application, it is important to investigate and **understand what is the alert about** (its nature, severity) and what are the **possible risks** it represents. By doing that, you can **find ways to fix** it or, at least, work around it if it's not possible to fix (very specific case in ones application).
- **Describe the differences in access management to view alerts for different security features**:
	Code scanning and dependabot alerts can be seem by anyone with `Write` access to the repository (I guess that makes sense, since they are going to fix it). Secret scanning need `Admin` access. 
> [!note]
> GitHub allows RBAC (role-based access control) so this is really very costumizable. What would be the best practices to manage access to security alerts? I think code scanning is interesting for all developers, but dependency are not always managed by everyone in the team. Sometimes, updating a dependency can affect the behavior of the application, so it should be reserved to tech leaders and above. What do you think?
- **Describe a security policy in a GitHub repository**:
	From what I understand, security policies and created in the `security.md` file, and basically are used to tell your contributors how to manage security vulnerabities and how to work secure within the repository. 
- **Identify where to use Dependabot alerts in the software development lifecycle**:
	GitHub Dependabot proactivelly analyses your dependency graph (not exactly the dependency graph, but both take information from the same place, so...) to find vulnerabilities. Anytime the **dependency graph changes** or a **new vulnerability is added** to GitHub's database, it **creates an alert**. It helps no only the development, but maintainence of your project in long term.  With GHAS, dependabot features are extended to dependency review, which is an workflow that **analyses new dependencies right in the pull request** and automatically block them if they are vulnerable.
> [!note]
> It goes further if you use **Version Updates** with dependabot using the `dependabot.yml`, where you can configure dependabot to automatically update your dependencies when there's a new version. It's pretty cool, but just like GitHub Actions, it's kinda complex. Has anyone here used it and want to talk about it?

**Reply by queenofcorgis** [reply]
Awesome job - thanks for making this compilation and sharing with us!

**Reply by yuca-1** [reply]
Yo diría que son más que abusivos y mentirosos que ellos mismos están haqueando cuentas haciéndose pasar por personas de inversiones pero yo diría que ser abusivos y envidiosos tarde o temprano se descubren yo creo que lo que debería Aser un reporte policiaco y así dar con este tipo de abusadores  envidiosos rasistas flojos haqueadores intimidadores acusadores etc .

**Comment by geopd** [comment]
1. B
2. D
3. C

**Comment by alexey-lebedev** [comment]
1. B
2. D
3. C

**Comment by davevad93** [comment]
That's great, thanks for the learning resources, looking forward for part 2 of the prep!

**Comment by keerthan0521** [comment]


**Reply by queenofcorgis** [reply]
@keerthan0521 

We’ve clarified our stance on using generative AI tools like ChatGPT within our Community via [this announcement](https://github.com/orgs/community/discussions/69006). Please review the guidelines to ensure your post meets them as failure to adhere to those rules can result in action taken by our moderator team.  You can read our updated [Code of Conduct](https://github.com/community/community/blob/baa5cd9020ef3f2bbf86dca1191d9c90f9b0822d/CODE_OF_CONDUCT.md#reasonable-use-of-ai-generated-content) and the [announcement](https://github.com/orgs/community/discussions/69006) for more details. Thank you for helping us maintain an authentic and beneficial space for everyone.

**Comment by queenofcorgis** [comment]
Pare Two's materials and prep questions are posted in a [new Discussion](https://github.com/orgs/community/discussions/138220) - see you there 🤓

**Comment by martinondiwa** [comment]
Great resources provided for the GitHub Advanced Security certification prep! I found the Microsoft Learn modules particularly useful for getting a solid understanding of GitHub Advanced Security's core features. For those preparing for the exam, I’d recommend also checking out the GitHub Docs for the latest updates and best practices. It’s always helpful to see real-world applications of these tools. Looking forward to seeing everyone's responses to the practice questions!

**Reply by mohamdmmaa78-maker** [reply]
السلام عليكم

**Comment by keerthan0521** [comment]
Thanks for the useful resources for GitHub advanced security certification. Resources I find very helpful for preparation are 1) Microsoft Learn modules 2) GHAS prep by Microsoft Press on LinkedIn Learning 3) Official GitHub documentation.

**Comment by jccampanero** [comment]
Following @queenofcorgis advice, these are my answers (I promise I didn't see the ones you provided!):

1. B
2. D
3. C

**Comment by trivikramm** [comment]
See the answers below
1. B
2. D
3. C

**Comment by profile-palash** [comment]
Question 1: Which of the following is a feature of Dependabot?

Answer: B. Dependabot opens pull requests to update dependencies.
<details>
<summary>Explanation</summary>
Dependabot helps keep your dependencies up-to-date by automatically opening pull requests to update them. This ensures that your project uses the latest versions, which can include important security patches and new features.
</details>
Question 2: Which of the following features is part of GitHub Advanced Security?

Answer: D. All of the above
<details>
<summary>Explanation</summary>
GitHub Advanced Security includes several features to enhance the security of your codebase:
- Secret Scanning: Detects and alerts on leaked secrets in your code.
- Dependabot: Creates alerts on and updates insecure dependencies.
- Code Scanning: Identifies and fixes security vulnerabilities and other errors in your code.
</details>
Question 3: Which of the following is a feature of GitHub Advanced Security?

Answer: C. Dependency scanning.
<details>
<summary>Explanation</summary>
Dependency scanning is a feature of GitHub Advanced Security that helps identify and manage vulnerabilities in your project's dependencies. This ensures that you are aware of and can address any security risks associated with the libraries and packages your project relies on.
</details>

**Comment by queenofcorgis** [comment]
We're back with another prep course, this time for the [GitHub Foundations Exam](https://github.com/orgs/community/discussions/154502)! We just kicked off [Week One of the prep course](https://github.com/orgs/community/discussions/155153) and we hope to have you studying alongside us ✏️ . [Start studying for Week One](https://github.com/orgs/community/discussions/155153) and for a chance to win one of our 25 certification exam vouchers.

**Comment by locnath56** [comment]
1  A
2 D
3 A

**Comment by mohamdmmaa78-maker** [comment]
> _This discussion and some of its comments have been edited and modified from its original format to enable evergreen learning_
> 
> Welcome to the first part of the [GitHub Advanced Security certification prep course](https://github.com/orgs/community/discussions/137197)! To begin, we’ll be going over some intro materials and best practices to build a strong foundation for the exam. This discussion is your space to ask questions, review study guides, and do some exam practice.
> 
> ## Step One: Prep 📚
> We’ve assembled some materials for this first section.
> 
> * Microsoft Learn Modules
>   
>   * [Introduction to GitHub Advanced Security](https://learn.microsoft.com/en-us/training/modules/introduction-github-advanced-security/?ns-enrollment-type=Collection&ns-enrollment-id=rqymc6yw8q5rey)
>   * [Configure Dependabot security updates on your GitHub repo ](https://learn.microsoft.com/en-us/training/modules/configure-dependabot-security-updates-on-github-repo/?ns-enrollment-type=Collection&ns-enrollment-id=rqymc6yw8q5rey)
> * Study guides from @NickLiffen
>   
>   * [Understanding GitHub Advanced Security](https://resources.github.com/learn/pathways/security/essentials/application-security-testing-github-advanced-security)
>   * [Enabling GitHub Advanced Security](https://resources.github.com/learn/pathways/security/essentials/enabling-github-advanced-security)
> * Study guides from @LadyKerr
>   
>   * [Describe the GHAS features and their functionality](https://github.com/LadyKerr/github-certification-guide/blob/main/study-guides/gh-advanced-security.md#domain-1-describe-the-ghas-security-features-and-functionality)
>   * [Configure and use Dependency Management ](https://github.com/LadyKerr/github-certification-guide/blob/main/study-guides/gh-advanced-security.md#domain-3-configure-and-use-dependency-management)
> * Videos
>   
>   * GitHub Advanced Security - Pass the Exam from freeCodeCamp
>     
>     * [GHAS Overview](https://www.youtube.com/watch?v=i740xlsqxEM&t=0s)
>     * [Dependency Management](https://www.youtube.com/watch?v=i740xlsqxEM&t=3677s)
> * Community member materials
>   
>   * @Salgado2004's [breakdown in their comment](https://github.com/orgs/community/discussions/137620#discussioncomment-10578129)
> 
> ## Step Two: Test Your Knowledge ⚡
> ### Question 1: Which of the following is a feature of Dependabot?
> A) Dependabot automatically commits changes to the main branch. B) Dependabot opens pull requests to update dependencies. C) Dependabot only works with public repositories. D) Dependabot can update Docker images directly.
> 
> ### Question 2: Which of the following features is part of GitHub Advanced Security?
> A) Secret Scanning to detect and alert on leaked secrets in your code. B) Dependabot to create alerts on and update insecure dependencies. C) Code Scanning to identify and fix security vulnerabilities and other errors in your code. D) All of the above
> 
> ### Question 3: Which of the following is a feature of GitHub Advanced Security?
> A) Real-time collaboration. B) Automated code formatting. C) Dependency scanning. D) Issue tracking.
> 
> _[View the answers in my comment](https://github.com/orgs/community/discussions/137620#discussioncomment-10572841) 🧠_
> 
> **Use the discussion below to share additional study resources and respond to our prep questions**
> 
> [Join Part Two's Discussion on Scanning](https://github.com/orgs/community/discussions/138220)

https://github.com/orgs/community/discussions/137620#discussion-7133327

**Comment by mohamdmmaa78-maker** [comment]
مصاري٥٠دولار

**Comment by brunocaetanobrito** [comment]
B
D
C

**Comment by Xpollya** [comment]
I passed my **GHAS certification exam** on my first attempt. Pass4surexams was part of my preparation, especially when I wanted some additional practice before the actual exam.
